/image-metadata

Vibecoder9000 opened the pull request:
This Pull Request introduces a metadata management system for background images, transitioning from a purely file-system-based approach to a synchronized JSON database.

This change lays the groundwork for future features such as background tagging and folder organization.

### 1. Metadata System to store metadata for every background image, co-located with the images themselves.
- **Metadata Fields:**
    - `hash`: SHA256 hash of the file, used to detect identical filenames
    - `aspectRatio`: Pre-calculated aspect ratio.
    - `dominantColor`: Average color calculated via Jimp (used for UI loading placeholders).
    - [isAnimated: Detection for GIFs, APNGs, and animated WebPs. (used to detect if the client needs to thumbnail)
    - `tags` & `folderIds`: Structure added for future folder organization features.
    - `addedTimestamp`: For sorting by time added instead of A-Z.
    - `thumbnailResolution`: Tracks the resolution thumbnails were generated at for cache invalidation.
- **Startup Sync:** Added syncBackgroundsMetadata to server-main.js. On startup, the server compares the file system against the JSON, adding new files, removing deleted ones, and regenerating metadata if resolution settings change.
- **Concurrency Control:** Uses `p-limit` library for controlled parallel processing of metadata generation.
- **Shared Constants:** Exported `ALLOWED_IMAGE_EXTENSIONS` and `BACKGROUNDS_METADATA_FILE` for use across modules.

### 2. Thumbnail & Image Processing Refactor (src/endpoints/thumbnails.js
- **Smart Caching:** Thumbnails now track the resolution they were generated at. If the user changes thumbnail quality/size settings in `config.yaml`, the system detects the mismatch and regenerates them.
- **Aspect Ratio Preservation:** Improved logic to calculate thumbnail dimensions based on the target pixel area while maintaining the original aspect ratio.
- **Optimization:** Animated files (GIF, APNG, Animated WebP) now skip expensive processing where appropriate.
- **Shared Constants:** `SKIPPED_EXTENSIONS` and `ALLOWED_IMAGE_EXTENSIONS` are now exported for reuse.

### 3. API Updates src/endpoints/backgrounds.js
- Refactored `/all`, `/upload`, `/delete`, and `/rename` endpoints.
- **Atomic Operations:** All metadata writes use `write-file-atomic` to prevent JSON corruption.
- **Uploads:** Now generate metadata and thumbnails immediately upon upload.
- **Renaming:** Ensures both the physical file, the thumbnail, and the metadata entry are updated synchronously with automatic conflict resolution.

### 4. Frontend
- Minor update to public/scripts/backgrounds.js to handle the JSON response from the upload endpoint (which now returns the full metadata object).

### 5. Dependencies
- **Added:** `image-size@^2.0.2` for fast dimension extraction without fully loading images into memory.
- **Added:** `p-limit@^6.2.0` for concurrency control during metadata generation.

## Checklist
- [x] I have read the [Contribution guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).
- [x] Tested background uploads (static and animated).
- [x] Tested background deletion and renaming.
- [x] Verified backgrounds/index.json generation on server startup.
- [x] Verified thumbnail regeneration on resolution changes.

Vibecoder9000 committed:
git push --force

Implement background management in json

lint

upload multer weirdness, fix for timestamp

Cohee1207 commented on default/content/index.json:
```diff
@@ -3,6 +3,10 @@
         "type": "settings"
     },
+    {
+        "filename": "backgrounds.json",
```
This shouldn't really qualify as a content file. I recommend generating it with the server code instead.

1. Its content is dynamic and is generated and managed by the backend, similarly to secrets.json
2. Content manager will not automatically repopulate it if it's deleted unless you delete the content log file first

Vibecoder9000 replied:
504c9f8

Cohee1207 commented on src/endpoints/backgrounds-manager.js:
```diff
@@ -0,0 +1,307 @@
+            addedTimestamp: Date.now(),
+        };
+    } catch (error) {
+        if (error.code === 'ENAMETOOLONG') {
```
Is this really a problem? Shouldn't other parts of the server already fail with this exact error?

Vibecoder9000 replied:
This was to fix a bug causing hanging on rename about 6 months ago. Worked at the time. deca3e6

Cohee1207 commented on src/endpoints/backgrounds-manager.js:
```diff
@@ -0,0 +1,307 @@
+        const backgroundsJsonPath = path.join(userDirectories.root, 'backgrounds.json');
+        const backgroundsFolderPath = userDirectories.backgrounds;
+        const thumbnailsBgPath = userDirectories.thumbnailsBg;
+        const ALLOWED_IMAGE_EXTENSIONS = new Set(['.png', '.jpg', '.jpeg', '.gif', '.webp', '.bmp', '.tif', '.tiff']);
```
Make it a shared constant.

Vibecoder9000 replied:
33b65c3

Cohee1207 commented on src/endpoints/backgrounds-manager.js:
```diff
@@ -0,0 +1,307 @@
+ */
+export async function syncBackgroundsMetadata(userDirectories) {
+    try {
+        const backgroundsJsonPath = path.join(userDirectories.root, 'backgrounds.json');
```
Make the file name a constant. also, why not put it into backgrounds directory as `index.json` instead?

Vibecoder9000 replied:
No reason in particular. 87e999d

Cohee1207 commented on src/endpoints/backgrounds-manager.js:
```diff
@@ -0,0 +1,307 @@
+        if (filesToProcess.length > 0) {
+            const totalFiles = filesToProcess.length;
+
+            for (let i = 0; i < totalFiles; i += CONCURRENCY_LIMIT) {
```
I'd consider using [p-limit](https://www.npmjs.com/package/p-limit) for this.
It's a fairly well known and popular concurrency manager library. Just use a version that supports Node 18.

Vibecoder9000 replied:
https://github.com/sindresorhus/p-limit/releases v6.2 works. febc43b

Cohee1207 commented on src/server-main.js:
```diff
@@ -266,7 +266,10 @@ async function preSetupTasks() {
-    await ensureThumbnailCache(directories);
+
+    for (const userDirectories of directories) {
+        await syncBackgroundsMetadata(userDirectories);
```
Other function accept `directories` and iterate them on their own.

Vibecoder9000 replied:
ac46d28

Cohee1207 commented on src/util.js:
```diff
@@ -106,6 +106,18 @@ export function getConfigValue(key, defaultValue = null, typeConverter = null) {
+ * Gets the configured thumbnail resolution.
+ * @returns {number} Thumbnail resolution (width * height)
+ */
+export function getThumbnailResolution() {
```
I'd rather put it in backgrounds.js module
Also the name doesn't reflect that it's  a bg  thumbnail  resolution specifically

Vibecoder9000 replied:
6ec8996

Vibecoder9000 committed:
remove duplicated logic to generate backgrounds.json

Vibecoder9000 committed:
remove duplicate logic to create backgrounds.js

Vibecoder9000 committed:
remove excessive enodent eror handling

Vibecoder9000 committed:
simplify shared extension constants

Vibecoder9000 committed:
move backgrounds.json to backgrounds/index.json

Vibecoder9000 committed:
use plimit

Vibecoder9000 committed:
move user directory iteration to syncBackgroundsMetadata insetad of server main

Vibecoder9000 committed:
create and move getBackgroundThumbnailResolution

Vibecoder9000 committed:
nitpick

Cohee1207 committed:
Merge branch 'staging' into metadata

Cohee1207 commented on src/endpoints/backgrounds-manager.js:
```diff
@@ -0,0 +1,311 @@
+/**
+ * Determines if a file should skip server-side thumbnail generation.
+ * @param {string} filename - The filename to check.
+ * @param {object} [imageMeta] - Optional metadata for the image.
```
Consider adding a typedef. Here's what Copilot generated, looks fine for me.
```js

/**
 * @typedef {Object} BackgroundImageMetadata
 * @property {string} hash - SHA-256 hash of the image file.
 * @property {number} aspectRatio - Aspect ratio (width / height) of the image.
 * @property {boolean} isAnimated - Whether the image is animated.
 * @property {string} dominantColor - Dominant color in hex format (e.g., '#RRGGBB').
 * @property {string[]} tags - Array of tags associated with the image.
 * @property {string[]} folderIds - Array of folder IDs the image belongs to.
 * @property {number} addedTimestamp - Timestamp when the image was added.
 * @property {number} [thumbnailResolution] - Optional thumbnail resolution (width * height).
 */

/**
 * @typedef {Object} BackgroundsMetadata
 * @property {number} version - Metadata version.
 * @property {Object.<string, BackgroundImageMetadata>} images - Mapping of filenames to their metadata.
 * @property {string[]} folders - Array of folder IDs.
 * @property {string[]} tags - Array of tags.
 */
```

Vibecoder9000 replied:
6bf1e3d

Cohee1207 replied:
Function param is still untyped.

Vibecoder9000 replied:
Whoops. 27250e3

Cohee1207 commented on src/endpoints/backgrounds-manager.js:
```diff
@@ -0,0 +1,311 @@
+function shouldSkipServerThumbnailGeneration(filename, imageMeta) {
+    const fileExtension = path.extname(filename).toLowerCase();
+    return SKIPPED_EXTENSIONS.includes(fileExtension) ||
+           (imageMeta && imageMeta.isAnimated) ||
```
```suggestion
           (imageMeta?.isAnimated) ||
```

Vibecoder9000 replied:
0483e2e

Cohee1207 commented on src/endpoints/backgrounds-manager.js:
```diff
@@ -0,0 +1,311 @@
+    const fileExtension = path.extname(filename).toLowerCase();
+    return SKIPPED_EXTENSIONS.includes(fileExtension) ||
+           (imageMeta && imageMeta.isAnimated) ||
+           ((fileExtension === '.png' || fileExtension === '.webp') && imageMeta && !imageMeta.thumbnailResolution);
```
```suggestion
           ((fileExtension === '.png' || fileExtension === '.webp') && !imageMeta?.thumbnailResolution);
```

Vibecoder9000 replied:
0483e2e

Cohee1207 commented on src/endpoints/backgrounds-manager.js:
```diff
@@ -0,0 +1,311 @@
+        dominantColor: hexColor,
+        tags: [],
+        folderIds: [],
+        addedTimestamp: Date.now(),
```
Shouldn't it be the file creation date to properly implement 'sort by added'?

Vibecoder9000 replied:
I'm not sure. I think date uploaded/seen by server would be more accurate? I'm guessing what users will upload.

Cohee1207 replied:
I think original creation dates should be preserved if you'd want to add sorting by create date.

Vibecoder9000 replied:
83c346f

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,280 @@
+     */
+    async update(updateFn) {
+        let metadata;
+        try {
```
Consider using `read` function here

Vibecoder9000 replied:
0571a19

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,280 @@
+            const rawData = await fsp.readFile(this.jsonPath, 'utf8');
+            return JSON.parse(rawData);
+        } catch (error) {
+            if (error.code === 'ENOENT' || error instanceof SyntaxError) {
```
How were these specific errors chosen? Any reason not to always regenerate the file?

Vibecoder9000 replied:
✨ 4d42e80

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,280 @@
 
-    const oldFileName = path.join(request.user.directories.backgrounds, sanitize(request.body.old_bg));
-    const newFileName = path.join(request.user.directories.backgrounds, sanitize(request.body.new_bg));
+        const fileExists = await fsp.access(filePath).then(() => true).catch(() => false);
```
There's `fileExists` below.

Vibecoder9000 replied:
28c80b0

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,280 @@
+            const data = metadata.images[oldFilename];
+            if (!data) {
+                const err = new Error(`Background '${oldFilename}' not found in metadata.`);
+                err.statusCode = 404;
```
These causes a type error. Either create a custom class for Error or just always use a 500 code.

Vibecoder9000 replied:
0cc15b5

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,280 @@
+ * @param {string} originalFilename - The original desired filename.
+ * @returns {Promise<string>} A unique filename.
+ */
+async function getUniqueFilename(directory, originalFilename) {
```
There's a `getUniqueName` util function that you can reuse

Vibecoder9000 replied:
b6c9c7f

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,280 @@
+
+/**
+ * Handles the upload of a new background image.
+ * @param {express.Request & { file: import('multer').File, user: any }} request - The Express request object.
```
You don't have to define a custom request type. Request types already include proper types both for File and user

Vibecoder9000 replied:
db76a87

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,280 @@
+            try { await fsp.unlink(finalBgPath); } catch { /* ignore */ }
+        }
+        if (request.file?.path) {
+            try { await fsp.unlink(request.file.path); } catch { /* ignore */ }
```
Is the upload not removed in happy path?

Vibecoder9000 replied:
In the happy path, the temporary file is moved (not copied) to its final destination using fsp.rename, so there's no temporary file left to clean up. The cleanup code with unlink calls is only executed when an error occurs during the upload process.

Cohee1207 commented on src/endpoints/thumbnails.js:
```diff
@@ -1,16 +1,21 @@
+export const publicRouter = express.Router();
+export const apiRouter = express.Router();
+
+export const SKIPPED_EXTENSIONS = ['.apng', '.mp4', '.webm', '.avi', '.mkv', '.flv', '.gif'];
```
You can also make this one a Set.

Vibecoder9000 replied:
962fd9e

Cohee1207 commented on src/endpoints/thumbnails.js:
```diff
@@ -91,162 +96,213 @@ export function invalidateThumbnail(directories, type, file) {
-    let originalFolder = getOriginalFolder(directories, type);
+export async function generateThumbnail(directories, type, file, forceGenerate = false, isKnownAnimated = null) {
+    // If the caller has already determined the file is animated, skip processing.
+    if (isKnownAnimated === true) {
```
No need for strict comparison with boolean literal.
```suggestion
    if (isKnownAnimated) {
```

Cohee1207 commented on src/endpoints/thumbnails.js:
```diff
@@ -91,162 +96,213 @@ export function invalidateThumbnail(directories, type, file) {
+        // If isKnownAnimated is false, we assume the caller knows it is static and skip this check.
+        if (fileExtension === '.webp' && isKnownAnimated !== false) {
+            const buffer = fs.readFileSync(pathToOriginalFile);
+            const isAnimatedWebP = buffer.indexOf('ANIM') !== -1 || buffer.indexOf('ANMF') !== -1;
```
Duplicated check from `generateSingleFileMetadata`?

Vibecoder9000 replied:
ad29602

Vibecoder9000 committed:
typdef

Vibecoder9000 committed:
simplify checks

Vibecoder9000 committed:
use read method

Vibecoder9000 committed:
error -> regenerate

Vibecoder9000 committed:
use fileExists

Cohee1207 commented on src/endpoints/backgrounds-manager.js:
```diff
@@ -0,0 +1,331 @@
+/**
+ * Generates all necessary metadata for a single background file.
+ * @param {string} filePath - The full path to the background file.
+ * @returns {Promise<object>} A metadata object. Throws an error if processing fails.
```
Add a returns type.

Vibecoder9000 replied:
dae4c0e

Vibecoder9000 committed:
type error fix (maybe)

Vibecoder9000 committed:
add object to jsdoc returns

Vibecoder9000 committed:
typo

Vibecoder9000 committed:
use getUniqueName from util.js

Vibecoder9000 committed:
remove duplicated multer something or other

Vibecoder9000 committed:
make skipped_extensions a set

Vibecoder9000 committed:
remove strict comparison

Co-authored-by: Cohee <18619528+Cohee1207@users.noreply.github.com>

Vibecoder9000 committed:
create isAnimatedWebP and use it

Vibecoder9000 committed:
Merge branch 'metadata' of https://github.com/Vibecoder9000/SillyTavern into metadata

Vibecoder9000 committed:
use file creation date instead of server date

Vibecoder9000 committed:
actually type BackgroundImageMetadata

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,263 @@
+            // Attempt to rename the thumbnail only if it exists.
+            await fsp.rename(oldThumbPath, newThumbPath);
+        } catch (thumbError) {
+            if (thumbError.code !== 'ENOENT') {
```
What if expecting an error, you check for thumbnail existence in advance?

Vibecoder9000 replied:
02aa7d2

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,263 @@
+        console.error(`Background upload failed for ${originalFilename}:`, err);
+
+        if (finalBgPath) {
+            try { await fsp.unlink(finalBgPath); } catch { /* ignore */ }
```
I'd prefer to use file exists check instead of try/catching.

Vibecoder9000 replied:
02aa7d2

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,263 @@
+            try { await fsp.unlink(request.file.path); } catch { /* ignore */ }
+        }
+
+        response.status(500);
```
This sets a status but doesn't complete the response. Use sendStatus(500)

Vibecoder9000 replied:
751d760

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,263 @@
-});
+router.post('/all', async function (request, response) {
+    try {
+        const manager = new BackgroundsMetadataManager(request.user.directories);
```
How does this ensure that the metadata is not stale? For example, if you directly drop images into backgrounds directory in runtime.

Vibecoder9000 replied:
f14b9da

Vibecoder9000 committed:
use fileExists instead of try catch

Vibecoder9000 committed:
response.sendStatus(500);

Vibecoder9000 committed:
sync json to allow adding images to the folder without restart

Vibecoder9000 committed:
lint

axAilotl commented:
I got background tabs (global, character, chat) and per-character background ready to go once this approved. I'll do a PR for each separately.

Cohee1207 commented:
I really have no idea how to proceed with this. Chat and character backgrounds would be excluded from receiving all the goodies the metadata json offers. I just don't see a big benefit in building such a complex system that has a very limited reach even in scope of its own domain area. Could we perhaps flip it around and consider the following:

- Make a generic endpoint that accepts a path to an image in the user data directory and returns a metadata object (with hash, dominant color, tags, etc.)
- Store cached info in some JSON and in-memory and use file timestamps to avoid reprocessing unless the file changes (how thumbnails work right now)
- Frontend will request this metadata for images that it needs to show, be it global, character, chat backgrounds, gallery items, whatever.

How does that sound @Vibecoder9000 @axAilotl ?

axAilotl commented:
Would be interesting, then can utilize the additional metadata in the future.

Per-Chat BG
/user/images/{{character}}/img.ext (from image-gen)
Per-Character BG
/characters/{{character}}/backgrounds/img.ext (from charx import, or other)
Sprites
/characters/{{character}}
Per-character Images
/user/images/{{character}}/* (from charx import, image gen, or other)

Vibecoder9000 committed:
Generic image metadata endpoint. Remove tags and starred. No startup sync.

Vibecoder9000 committed:
lint

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,311 @@
+/**
+ * POST /folders - Update virtual folders.
+ */
+router.post('/folders', async function (request, response) {
```
Wait, folders? How do I test that?

Vibecoder9000 replied:
Oops, got a little too ambitious adding the route. Folders is a later PR. 701accc

Cohee1207 commented on src/endpoints/backgrounds.js:
```diff
@@ -1,76 +1,311 @@
+ * @param {string} filePath - The full path to the file.
+ * @returns {Promise<boolean>} True if the file exists, false otherwise.
+ */
+async function fileExists(filePath) {
```
Move it to utils module. `getUniqueFilename` too.

Vibecoder9000 replied:
25ab6e2

Cohee1207 commented on src/endpoints/image-metadata.js:
```diff
@@ -0,0 +1,282 @@
+import { Jimp } from '../jimp.js';
+import { getConfigValue } from '../util.js';
+
+export const METADATA_FILE = 'index.json';
```
The name is not descriptive enough if it's supposed to be in the data root folder. `images.json`? or `image-metadata.json`?

Vibecoder9000 replied:
25ab6e2

Cohee1207 replied:
I misunderstood this. Thought it was per-user but it's actually per-directory. Then `index.json` is fine. On the other hand, chat backgrounds again gain no love and still unindexed.

Vibecoder9000 replied:
71bd28a

Vibecoder9000 committed:
Descriptive json and move to util.js

Vibecoder9000 committed:
remove folders routes

Cohee1207 committed:
It's a POST

Cohee1207 committed:
Merge branch 'staging' into metadata

Cohee1207 committed:
Revert file name to index.json

Cohee1207 committed:
Type errors fixing

Cohee1207 committed:
Remove p-limit

Cohee1207 committed:
Why?

Cohee1207 commented:
@codex review

Cohee1207 commented:
@Vibecoder9000 how would you look at that if I take out all the changes to existing endpoints and add a separate endpoint for requesting image metadata by path(s)?

Vibecoder9000 commented:
I'm good with really any way to integrate it.

Vibecoder9000 committed:
apply codex suggestions

Vibecoder9000 committed:
Merge branch 'metadata' of https://github.com/Vibecoder9000/SillyTavern into metadata

Vibecoder9000 commented on src/endpoints/thumbnails.js:
```diff
@@ -91,162 +97,213 @@ export function invalidateThumbnail(directories, type, file) {
+        const file = sanitize(rawFile);
+        if (file !== rawFile) return response.sendStatus(403);
 
-        if (!thumbnailsEnabled) {
+        const serveOriginal = () => {
             const folder = getOriginalFolder(request.user.directories, type);
+            const pathToOriginalFile = path.resolve(path.join(folder, file));
+            if (!fs.existsSync(pathToOriginalFile)) return response.sendStatus(404);
+            return response.sendFile(pathToOriginalFile);
```
[99994a8](https://github.com/SillyTavern/SillyTavern/pull/4788/commits/99994a8169c673a6e4741e6dc2f15f514b718dc5)

Vibecoder9000 commented on src/endpoints/thumbnails.js:
```diff
@@ -1,16 +1,22 @@
+export const apiRouter = express.Router();
+
+export const SKIPPED_EXTENSIONS = new Set(['.apng', '.mp4', '.webm', '.avi', '.mkv', '.flv', '.gif']);
+export const ALLOWED_IMAGE_EXTENSIONS = new Set(['.png', '.jpg', '.jpeg', '.gif', '.webp', '.bmp', '.tif', '.tiff']);
```
[99994a8](https://github.com/SillyTavern/SillyTavern/pull/4788/commits/99994a8169c673a6e4741e6dc2f15f514b718dc5)

Vibecoder9000 committed:
see comment

Vibecoder9000 commented:
## Summary of Changes

1. Reverted  `backgrounds.js` to original simple form
   - Removed all metadata integration  
   - Restored original `getImages` function usage
   - Restored simple CRUD operations (delete, rename, upload)  
   - Upload now returns plain text filename instead of JSON  

2. Reverted `public/scripts/backgrounds.js`  
   - Changed upload response handling back to `response.text()` instead of `response.json()`  

3. Updated `image-metadata.js` to be a standalone endpoint  
   - Added Express router with `POST /api/image-metadata` endpoint  
   - Accepts `path` (single) or `paths` (array) in request body  
   - Paths are relative to user data directory  
   - Returns metadata objects with on-demand generation and mtime-based caching  
   - Includes security check to ensure paths are within user data directory  

4. Registered new endpoint in `server-startup.js`  
   - Added import for `imageMetadataRouter`  
   - Registered at `/api/image-metadata`  

---

## New API Endpoint

**POST** `/api/image-metadata`

### Request body:

```json
// Single path
{ "path": "backgrounds/my-image.png" }

// Multiple paths
{ "paths": ["backgrounds/img1.png", "characters/avatar.jpg"] }
````

### Response:

Single path: Returns metadata object directly
Multiple paths: Returns object mapping paths to metadata

Example response:

```json
{
  "hash": "abc123...",
  "aspectRatio": 1.7778,
  "isAnimated": false,
  "dominantColor": "#4a5568",
  "folderIds": [],
  "addedTimestamp": 1704567890123,
  "thumbnailResolution": 14400
}
```

Cohee1207 commented:
Cool, that's exactly what I've been looking for. I'll review in a few (unit unspecified).

Wolfsblvt commented:
Can we give this PR a meaningful name? Might be good for future clean commit history, and also for notifcations I receive about this PR :D

Cohee1207 commented on src/endpoints/thumbnails.js:
```diff
@@ -91,162 +97,214 @@ export function invalidateThumbnail(directories, type, file) {
+                    const dimensions = sizeOf(buffer);
+                    const ratio = (dimensions.height > 0) ? (dimensions.width / dimensions.height) : 1.0;
+                    // When a thumbnail exists, return the current resolution from config so the JSON can be updated.
+                    const resolution = getBackgroundThumbnailResolution();
```
Is this resolution always going to be used, even if the thumbnail type is not 'bg'?

Vibecoder9000 replied:
Nope. c6c6883

Cohee1207 commented on src/endpoints/thumbnails.js:
```diff
@@ -1,16 +1,22 @@
+export const SKIPPED_EXTENSIONS = new Set(['.apng', '.mp4', '.webm', '.avi', '.mkv', '.flv', '.gif']);
+export const ALLOWED_IMAGE_EXTENSIONS = new Set(['.png', '.jpg', '.jpeg', '.gif', '.webp', '.bmp', '.tif', '.tiff', '.apng']);
 
-const thumbnailsEnabled = !!getConfigValue('thumbnails.enabled', true, 'boolean');
```
Why did we move it from here into a local scope of the endpoint?

Vibecoder9000 replied:
<img width="500" height="284" alt="image" src="https://github.com/user-attachments/assets/a281f3fc-63b3-4dd6-9d08-76ec8eb7977f" />

1cd4bd0

Cohee1207 commented on src/util.js:
```diff
@@ -594,6 +595,40 @@ export function getUniqueName(name, exists) {
+ * @param {string} originalFilename - The original desired filename.
+ * @returns {Promise<string>} A unique filename.
+ */
+export async function getUniqueFilename(directory, originalFilename) {
```
Don't think it's used anymore?

Vibecoder9000 replied:
e286dbf

Cohee1207 commented on src/util.js:
```diff
@@ -594,6 +595,40 @@ export function getUniqueName(name, exists) {
+ * @param {string} filePath - The full path to the file.
+ * @returns {Promise<boolean>} True if the file exists, false otherwise.
+ */
+export async function fileExists(filePath) {
```
Same here. Looks unused after a clean-up.

Vibecoder9000 replied:
e286dbf

Cohee1207 commented on src/endpoints/image-metadata.js:
```diff
@@ -0,0 +1,386 @@
+        // Helper to resolve and validate a path
+        const resolvePath = (relativePath) => {
+            // Normalize path separators
+            const normalizedPath = relativePath.replace(/\\/g, '/');
```
Shouldn't there be a node method that does the same? Like 
```js
import posixPath from 'node:path/posix';
posixPath.normalize();
```

Vibecoder9000 replied:
4cd3d92

Cohee1207 commented on src/endpoints/image-metadata.js:
```diff
@@ -0,0 +1,386 @@
+
+            // Group paths by directory for efficient batch processing
+            /** @type {Map<string, {relativePath: string, filename: string}[]>} */
+            const pathsByFolder = new Map();
```
My honest opinion: use a single metadata index file for the entire user data directory?

Why? It would be easier to maintain and clean-up: just iterate over a single file and check for existance of all referenced files. Otherwise, I'm afraid it may leave traces in some unexpected places like images for deleted characters that may not be found by a data clean-up dialog.

Vibecoder9000 replied:
Hopefully addressed with 71bd28a to support different image types.

Cohee1207 reviewed (changes requested):
We're almost there. Just a couple of points to clarify.

Vibecoder9000 committed:
remove unused functions

Vibecoder9000 committed:
remove path normalization for da path, not needed

Vibecoder9000 committed:
Fix dimensions shadowing issue, circular imports, and apply bg res to only bg

Vibecoder9000 committed:
move thumbnailsEnabled to the top level

Vibecoder9000 committed:
Merge branch 'staging' of https://github.com/SillyTavern/SillyTavern into metadata

Vibecoder9000 committed:
Metadata now stored at {userDataRoot}/image-metadata.json instead of per-folder index.json
Uses relative paths as keys (e.g., backgrounds/sunset.png) for easy file existence checking
Added cleanupOrphanedMetadata() function and /api/image-metadata/cleanup endpoint for data-maid integration
Simplified the API endpoint by removing folder-grouping logic

Vibecoder9000 committed:
lint

Vibecoder9000 committed:
update backgrounds.js imports

Vibecoder9000 committed:
Merge branch 'staging' into metadata

Cohee1207 commented on src/endpoints/image-metadata.js:
```diff
@@ -0,0 +1,419 @@
+        dominantColor,
+        folderIds: [],
+        addedTimestamp,
+        thumbnailResolution: getThumbnailResolution('bg'),
```
Should it be hard coded to 'bg'? Also, how does it react to the config being changed?

Vibecoder9000 replied:
Yeah, I forgot about that after changing the file to support other image types. It reacts to config changes after a restart. I don't think it'll be changed enough to support immediate thumbnailing. 
805c50a

Cohee1207 commented on src/endpoints/image-metadata.js:
```diff
@@ -0,0 +1,419 @@
+ * @param {string} relativePath - The relative path to remove
+ */
+export async function removeMetadata(userDataRoot, relativePath) {
+    const normalizedPath = relativePath.split(path.sep).join('/');
```
This could look more readable
`const posixPath = relativePath.replaceAll(path.sep, path.posix.sep)`

Vibecoder9000 replied:
a15c3c7

Cohee1207 commented on src/endpoints/image-metadata.js:
```diff
@@ -0,0 +1,419 @@
+    const orphanedPaths = [];
+
+    for (const relativePath of Object.keys(index.images)) {
+        const fullPath = path.join(userDataRoot, relativePath);
```
I think you should also check for `if (!isPathUnderParent(userDataRoot, fullPath)) {` here to avoid injected out of data root paths.

Vibecoder9000 replied:
a15c3c7

Vibecoder9000 committed:
security check, .replaceAll

Vibecoder9000 committed:
Remove useless function, pass type to cache invalidate and seperate file types

Cohee1207 committed:
Merge branch 'staging' into metadata

Cohee1207 committed:
fix: normalize file paths to use POSIX separators in metadata functions

Cohee1207 reviewed (approved):
Looks good.

Cohee1207 commented:
@codex review

Cohee1207 commented on src/endpoints/thumbnails.js:
```diff
@@ -91,162 +92,212 @@ export function invalidateThumbnail(directories, type, file) {
 
-        if (!pathToCachedFile) {
-            return response.sendStatus(404);
+        // Serve original for animated formats or GIFs
+        if (animatedEnabled && isAnimatedFormat) {
+            return serveOriginal();
```
> // If generation failed (path is null), send 404 for frontend placeholder

Replace with `serveOriginal()`? I don't see any other way to address this.

Vibecoder9000 replied:
I'm having a hard time actually understanding the explanations, but claude seems to think that's the best approach too. 4c47398

Vibecoder9000 committed:
serve original

Vibecoder9000 commented:
<img width="1200" height="740" alt="image" src="https://github.com/user-attachments/assets/db385e9c-a70a-4131-be1a-7ec59621e7bc" />
